### DISCUSSION TITLE
Is it possible to use Windmill behind an HTTP1/1 proxy?

### DISCUSSION THREAD

**Discussion by bigcat88** [windmill-labs/windmill#3747] [Questions] [answered]
I am reaching out to seek advice regarding a potential integration of Windmill as an external application within Nextcloud.

Our current app system in Nextcloud has certain limitations. Specifically, all external applications, unless their UI is modified (which we aim to avoid), must operate within an iframe, with the frontend connecting to the backend via a proxy. The current version of the proxy does not support gRPC, websockets, or any persistent connections. It only supports stateless requests such as REST API, i.e., HTTP/1.1.

Would these limitations be critical for Windmill? From my brief review of the repository, I understand that collaborative flow editing (which we do not need) and LSP would not work. Additionally, creating Python scripts would not be possible as their editing uses websockets.

I have also noticed that gRPC is used in some parts of Windmill.

Therefore, my questions are:

Will there be more components in Windmill in the future that rely on protocols other than HTTP/1.1?
If not, would it be technically feasible for us to package Windmill into a single Docker container (including the database, server, and workers based on the number of CPU cores where the container is deployed) and use it?

Thank you in advance for your assistance.

**Comment by rubenfiszel** [comment, marked as answer]
Windmill doesn't use grpc and websockets are really optional, only use for multiplayer and LSP.

You can run windmill in MODE=standalone where it will act as both a server and a worker. You will need a database though.

For the time being, we will keep being usable as a pure http service

**Comment by bigcat88** [comment]
Is it possible in the standalone container
 (`docker run -it -e DATABASE_URL=... ghcr.io/windmill-labs/windmill:1.338`) - to modify the container so that it spawned more than one worker inside a container? @rubenfiszel 
 
 If there is no such config option, can you please point in which file this logic lives and I'll start to learning Rust to patch this 😂
 
 Thanks a lot in advice

**Reply by rubenfiszel** [reply]
You can if you set NUM_WORKERS=X but it's not recommended for many reasons. One because tokio does not handle well having many command interrupts, and second because isolation is not guaranteed. But you do you
